generators/chipyard/src/main/resources/csrc/spiketile.cc

```diff
@@ -0,0 +1,917 @@
+                           unsigned char dcache_d_grantack,
+                           long long int dcache_d_sourceid,
+                           long long int dcache_d_data_0,
+                           long long int dcache_d_data_1,
+                           long long int dcache_d_data_2,
+                           long long int dcache_d_data_3,
+                           long long int dcache_d_data_4,
+                           long long int dcache_d_data_5,
+                           long long int dcache_d_data_6,
+                           long long int dcache_d_data_7,
+
+                           unsigned char mmio_a_ready,
+                           unsigned char* mmio_a_valid,
+                           long long int* mmio_a_address,
+                           long long int* mmio_a_data,
+                           unsigned char* mmio_a_store,
+                           int* mmio_a_size,
+
+                           unsigned char mmio_d_valid,
+                           long long int mmio_d_data
+                           )
+{
+  if (!host) {
+    host = context_t::current();
+    sout.rdbuf(std::cerr.rdbuf());
+    log_file = new log_file_t(nullptr);
+  }
+  if (tiles.find(hartid) == tiles.end()) {
+    printf("Constructing spike processor_t\n");
+    isa_parser_t *isa_parser = new isa_parser_t(isa, "MSU");
+    chipyard_simif_t* simif = new chipyard_simif_t(icache_ways, icache_sets,
+                                                   dcache_ways, dcache_sets,
+                                                   cacheable, uncacheable, readonly_uncacheable, executable,
+                                                   icache_sourceids, dcache_sourceids);
+    processor_t* p = new processor_t(isa_parser,
+                                     "vlen:128,elen:64",
+                                     simif,
+                                     hartid,
+                                     false,
+                                     endianness_little,
+                                     false,
+                                     log_file->get(),
+                                     sout);
+
+    p->enable_log_commits();
+
+    s_vpi_vlog_info vinfo;
+    if (!vpi_get_vlog_info(&vinfo))
+      abort();
+    for (int i = 1; i < vinfo.argc; i++) {
+      std::string arg(vinfo.argv[i]);
+      if (arg == "+spike-debug") {
+        p->set_debug(true);
+      }
+      if (arg == "+spike-stq") {
+        simif->use_stq = true;
+      }
+    }
+
+    p->reset();
+    p->get_state()->pc = reset_vector;
+    tiles[hartid] = new tile_t(p, simif);
+    printf("Done constructing spike processor\n");
+  }
+  tile_t* tile = tiles[hartid];
+  chipyard_simif_t* simif = tile->simif;
+  processor_t* proc = tile->proc;
+
+  simif->cycle = cycle;
+  if (debug) {
+    proc->halt_request = proc->HR_REGULAR;
+  }
+
+  proc->get_state()->mip->backdoor_write_with_mask(MIP_MTIP, mtip ? MIP_MTIP : 0);
```

Comment:
How about other CSRs? Like time/cycle etc

Reply:
`mcycle` is written in `spike_thread_main`

Reply:
`mtime` is accessed over the bus.

Reply:
I guess the question is does mcycle sync across tiles. For example, if you have multiple spike tiles or 1 spike + 1 rocket tile

Reply:
I believe mcycle only increments when the harts are out of reset. So its not guaranteed that they are in sync.

The current implementation does keep a consistent mcycle. The counter is tracked in the chisel and passed to the model.

Reply:
Thanks for the explanation. That makes sense.